src/new_client/client.rs

```diff
@@ -0,0 +1,223 @@
+                            let _ = messages.unbounded_send(Message {
+                                subject: subject.clone(),
+                                reply: reply_to.clone(),
+                                data: payload.clone(),
+                                responder: None,
+                            });
+                        }
+                    }
+
+                    ServerOp::Err(msg) => {
+                        log::error!("received -ERR '{}'", msg);
+                    }
+
+                    ServerOp::Unknown(line) => {
+                        log::warn!("unknown op: {}", line);
+                    }
+                }
+            }
+
+            // The user has enqueued an operation.
+            msg = user_ops.next().fuse() => {
+                match msg.expect("user_ops disconnected") {
+                    UserOp::Pub { subject, reply_to, payload } => {
+                        // Send a PUB operation to the server.
+                        encode(&mut writer, ClientOp::Pub {
+                            subject,
+                            reply_to: None,
+                            payload
+                        })
+                        .await?;
+                    }
+
+                    UserOp::Sub { subject, queue_group, sid, messages } => {
+                        // Add the subscription to the list.
+                        subscriptions.push((subject.clone(), sid, messages));
+
+                        // Send a SUB operation to the server.
+                        encode(&mut writer, ClientOp::Sub {
+                            subject,
+                            queue_group,
+                            sid,
+                        })
+                        .await?;
+                    }
+
+                    UserOp::Unsub { sid, max_msgs } => {
+                        // Remove the subscription from the list.
+                        subscriptions.retain(|(_, s, _)| *s != sid);
+
+                        // Send an UNSUB operation to the server.
+                        encode(&mut writer, ClientOp::Unsub {
+                            sid,
+                            max_msgs,
+                        })
+                        .await?;
+                    }
+
+                    UserOp::Ping { pong } => {
+                        // Send a PING operation to the server.
+                        encode(&mut writer, ClientOp::Ping).await?;
+
+                        // Record that we're expecting a pong.
+                        pongs.push_back(pong);
+                    }
+
+                    UserOp::Close => {
+                        // TODO(stjepang): Perhaps we should flush before closing abruptly.
+                        return Ok(());
+                    }
+                }
+            }
+
+            // Periodically flush writes to the server.
+            _ = Timer::at(next_flush).fuse() => {
+                writer.flush().await?;
```

Comment:
How does it behave if it can't flush whole buffer? If it blocks instead of returning success on partial flush, this operation can deadlock. If recently flushed command make server to send huge amount of data back, it might end up blocking on on it's own writes, waiting for client to read data from TCP stream. 

Timeout on flush can help to resolve deadlock

Reply:
this is still a draft. deadlock doesn't happen in this system because of the nats-server's flush logic that ejects slow consumers. duplex progress is incoming.